File: java/sort/linked_list/BubbleSort.java
package sort.linked_list;

public class BubbleSort {
    public void sort(ListNode head) {
        ListNode border = null;
        while (head.next.next != border) {
            boolean flag = false;
            ListNode prev = head;
            ListNode current = head.next;
            while (current.next != border) {
                if (current.value > current.next.value) {
                    ListNode next = current.next;
                    current.next = next.next;
                    next.next = prev.next;
                    prev.next = next;

                    flag = true;
                    current = prev.next;
                }
                prev = prev.next;
                current = current.next;
            }
            border = current;

            if (!flag) {
                break;
            }
        }
    }

    public class ListNode {
        public int value = 0;
        public ListNode next = null;

        public ListNode(int value) {
            this.value = value;
        }
    }
}

File: py/sort/linked_list/BubbleSort.py
#!/usr/bin/env python
# -*- coding:gbk -*-

class ListNode:
    def __init__(self, value: int):
        self.value, self.next = value, None

class BubbleSort:
    def sort(self, head: ListNode) -> None:
        border = None
        while head.next.next != border:
            flag, prev, current = False, head, head.next
            while current.next != border:
                if current.value > current.next.value:
                    next = current.next
                    current.next, next.next, prev.next = next.next, prev.next, next

                    flag, current = True, prev.next
                prev, current = prev.next, current.next
            border = current

            if not flag:
                break
